[2단계 - 자동차 경주 리팩터링] 오찌(오지훈) 미션 제출합니다.

Ohzzi opened the pull request:
2단계 미션으로 돌아왔습니다. 항상 좋은 피드백 감사합니다 제이! :)

1단계 미션을 merge할 때 피드백을 받고 나서, 2단계 미션을 진행하면서 가장 고민했던 부분은, 제가 MVC 패턴이라고 짜놓고는 정작 Model, View, Controller의 책임을 전혀 이해하지 못하고 있는 것이 아닌가 하는 생각이었습니다! 저는 처음에 Controller에 멤버 변수가 있는 것이 무엇이 문제가 되지? 하고 생각했으니까요.

이 부분은 저희 데일리 미팅 조 크루들과 따로 이야기를 나누어보면서 정리가 된 것 같습니다.
1. 컨트롤러에 Cars와 TryCount를 멤버 변수로 주는 것은 컨트롤러에 해당 객체에 대한 책임을 주는 것이다.
2. 컨트롤러가 모델의 상태를 가지는 것은 올바른 역할이 아니다.
모델과 뷰는 서로의 상태를 모르고, 컨트롤러는 이들을 이어주는 역할만 해야 한다.
3. 컨트롤러가 상태를 가지게 되면 동시성 이슈가 생길 수 있다.
    - 찾아보다보니 이와 비슷하게 스프링에서 컨트롤러를 포함한 Bean들은 Bean의 스코프를 변경하지 않는 이상 싱글톤이라는 점,
이로 인해 절대 상태를 갖지 않도록 해야 한다는 것을 알 수 있었습니다.

그래서 멤버 변수를 가져서 상태를 가지도록 하지 않고, 메소드를 호출할 때 지역 변수로 선언해 계속 다른 메소드를 호출할 때 인자로 넘겨주는 식으로 수정했습니다.

그리고! 1단계 두 번째 리뷰로 남겨주신 다른 질문들에 대한 대답은 1단계 PR(https://github.com/woowacourse/java-racingcar/pull/281) 각각의 코멘트에 남겨두었습니다 ;)

반영하지는 않았지만 몇 가지 고민한 부분이 있었다면

1. TryCount를 아예 불변 객체로 사용할 것인가?
    - 지금은 RacingController의 race() 메소드가 총 시도 횟수를 인자로 받아온 뒤에 int 형의 currentTryCount를 선언해서 해당 값을 증가시키면서 턴을 진행합니다. 비교는 TryCount 객체의 isNotSame() 메소드를 이용하고요.
    - 하지만 currentTryCount도 TryCount로 만들고, TryCount 값이 증가할 때 새 객체를 만들어서 던져주는 형태로 불변 객체로 만들어주면 나중에 저나 다른 사람이 코드를 작업할 때 race() 메소드 내에서 currentTryCount를 의도치 않게 조작하는 상황이나 동시성 문제로 인한 값 오염을 차단할 수 있지 않을까 하는 생각이 들기도 합니다.
    - 그런데 또 생각해보면 어차피 컨트롤러에 멤버 변수를 없애서 동시성 문제를 차단해 놓은 상태인데 굳이? 라는 생각이 들기도 해서요, 이 부분에 대해서 제이의 의견이 궁금합니다. 물론 저도 불변 객체의 사용과 그 장단에 대한 내용을 더 공부해서 결론을 내려보도록 하겠습니다 :)

2. 지난번 피드백 당시의 UnModifiableList와 비슷한 질문인데, 안전성을 위해서 인자로 들어가는 데이터가 불변해야 한다면 final을 붙이시는 것을 선호하는 편인가요? 저는 여태까지는 붙이지 않았습니다만, 저도 모르게 값을 바꾸거나 하는 상황을 방지하려면 final을 일일이 붙여주는 것이 좋겠다는 생각이 들어서요, 이 부분에 대해서도 제이의 의견이 궁금합니다!

Ohzzi committed:
refactor: position 비교 시 getter 대신 Position 클래스 안에서 boolean으로 반환하도록 수정

Ohzzi committed:
test: 분리한 도메인에 대한 테스트 추가

Ohzzi committed:
test: @SuppressWarning 대신 @DisplayName 사용하도록 수정

Ohzzi committed:
refactor: format 사용 방식 수정

format과 join으로 만든 문자열 뒤에 +로 다른 문자열을 붙이지 않고 format만으로 처리

Ohzzi committed:
refactor: Cars 객체를 생성할 때 배열을 만들지 않고 List와 stream을 사용하는 방식으로 통일

Ohzzi committed:
refactor: CarStatus의 이름을 명확하게 하기 위해 CarInfo로 변경

Ohzzi committed:
refactor: RacingController에서 멤버 변수를 사용하지 않도록 수정

컨트롤러에서 멤버 변수를 사용하게 될 경우 여러 사용자가 동시에 게임을 진행했을 때 멤버 변수의 공유 문제가 발생할 수 있어 좋은 방법이 아님. 따라서 메소드에서 객체를 만들어서 계속 인수로 던지는 형태로 리팩토링.

Ohzzi committed:
refactor: validate 과정에서 인자를 넘겨주지 않도록 수정

Ohzzi committed:
refactor: 잘못된 메소드 명 수정

Ohzzi committed:
refactor: 불필요한 toString 제거

jjj0611 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -12,41 +12,43 @@
         } catch (IllegalArgumentException | IOException e) {
             OutputView.printException(e);
-            start();
+            return inputCars();
```
입력에 대한 재귀도 잘 처리해주셨네요 :)

jjj0611 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -12,41 +12,43 @@
         } catch (IllegalArgumentException | IOException e) {
             OutputView.printException(e);
-            inputTryCount();
+            return inputTryCount();
```
ditto

jjj0611 commented on src/main/java/racingcar/view/OutputView.java:
```diff
@@ -15,9 +15,9 @@ public static void printStartMessage() {
-            System.out.println(createStatusString(carStatus));
+    public static void printCurrentCarInfo(List<CarInfo> carInfos) {
+        carInfos.forEach(carInfo -> {
+            System.out.println(createCarInfoString(carInfo));
```
초기에 toString을 쓰는 것보다 조금 복잡해졌을지 몰라도(객체의 수가 증가했음) car 객체에서 view의 로직을 분리해내니 각 객체가 하는 역할이 더 깔끔해졌죠? view 또한 '로직'이 있습니다. 다만 view는 view의 로직을 도메인은 도메인 로직을 담당해야 합니다 :)

아직 어떤 것이 도메인 로직이고, 어떤 것이 view인지가 애매할 수 있겠지만, 차차 감을 잡아가실 거에요! :)

jjj0611 commented on src/main/java/racingcar/model/car/Cars.java:
```diff
@@ -27,20 +27,20 @@ public void moveAll(RandomNumberGenerator random) {
-                .map(CarStatus::new)
+                .map(CarInfo::new)
                 .collect(Collectors.toUnmodifiableList());
     }
```
오찌 CarInfo의 네이밍에 대한 부분은 1단계 답변 주신 것에 대하여 남겼어요.
그런데 Cars라는 객체에서 CarInfo를 반환해야하는게 맞을지 고민해보면 좋겠어요 :)
MVC 패턴에서 레이어를 잘 구분하려면 어떤 부분이 필요할지 고민해보죠!

jjj0611 commented on src/main/java/racingcar/model/car/Winners.java:
```diff
@@ -11,7 +11,7 @@
         int maxPosition = getMaxPosition(cars);
         names = cars.stream()
-                .filter(car -> car.isMaxPosition(maxPosition))
+                .filter(car -> car.isSameWithPosition(maxPosition))
```
저는 보통 이럴 때 isSameWith(maxPosition)정도로 표현하는 편이에요!

jjj0611 reviewed (changes requested):
안녕하세요 오찌!

1단계 코멘트에 답글 남겨두었으니 한 번 확인 부탁드려요 :)

- TryCount를 아예 불변 객체로 사용할 것인가?

이건 좀 더 공부하고 대립을 해본 후에 말씀드려도 재밌을 것 같네요!
불변으로도 만들어보시고, 아닌 상태로도 구현해보시면 어떨까요?

- final 키워드에 대한 고민

 저는 보통 붙이지 않습니다. 메서드 파라미터에 final을 붙인다는건 파라미터로 들어온 친구에 재할당하지 않겠다는 건데, 보통 메서드 라인 내에서 재할당하는 경우는 거의 없다고 생각해요. 또한 메서드 라인을 짧게 가져간다면 재할당을 한다고 하더라도 쉽게 찾을 수 있구요. 저는 그래서 파라미터에는 거의 붙이지 않는 편입니다.

오찌 2단계도 고생 많았습니다!
1단계 때 그래도 반영된 구조가 조금 있어서 엄청 큰 변경 사항은 없었네요.
1단계 때 답글 달아드린 것도 한 번 확인해주시고 질문이 있으시다면 2단계 PR에 같이 녹여서 이야기 해주세요.
간단히 고민해볼만한 부분 남겼으니 확인해주시고, 코스 과정 내 자동차 경주 피드백에 들어 있는 부분들을 적용해봐도 좋을 것 같아요 :)

1단계 마지막 리뷰에 답글 남겨주신 것에 대해 이야기 한 부분들도 고민해보시고 같이 녹여주세요!

Ohzzi committed:
refactor: Car의 DTO 클래스 이름 변경

Ohzzi committed:
refactor: 모든 필드와 메소드가 static인 클래스의 생성자를 private으로 막도록 수정

Ohzzi committed:
refactor: Car를 DTO로 변환하는 위치를 Controller 단으로 이동

Ohzzi committed:
refactor: 클래스명 변경에 따른 메소드명 수정

Ohzzi committed:
refactor: Controller에서 게임 진행 여부 판단을 위한 시도 횟수 비교 시 int가 아닌 불변 객체 TryCount를 사용하도록 수정

Ohzzi committed:
refactor: Position 클래스도 TryCount처럼 불변 객체가 되도록 수정

Ohzzi committed:
test: 테스트 오류 수정

Ohzzi committed:
refactor: 불필요한 메소드 삭제하고 Comparalbe의 raw한 사용 수정

Ohzzi committed:
refactor: 불필요한 map 제거

Ohzzi commented:
피드백 내용도 반영하고 개인적으로 몇 가지 고민사항들을 적용해보고 다시 리뷰 요청을 드립니다!!
아래는 이번에 피드백 반영 이후 중점적으로 리팩토링한 부분입니다!

---
### DTO 생성 위치 변경
이전에는 Cars 객체에서 DTO를 생성했었는데요, Cars도 도메인 계층에 들어가는 클래스인데 마치 컨트롤러나 서비스인 것 처럼 사용하고 있었던 것 같습니다! 찾아보니까 도메인에서는 DTO를 사용하지 않는 것 같더라고요! 그래서 한 단계 위로 올려서 컨트롤러에서 DTO를 생성하도록 해주었습니다. 다만 이럴 경우에 어쩔 수 없이 Cars에 getter를 추가해서 사용하게 되어서 그 부분이 조금 걸리네요.

### TryCount, Position 사용 방식 변경
1단계 리뷰때 제가 고민중이라고 했던 방식대로, TryCount와 Position 모두 값을 변경시킬 시 새로운 객체를 생성해서 반환하도록 하는 불변 객체로 만들어주었습니다. 맨 처음에 이런 생각을 하게 된 것은, 컨트롤러에서 TryCount와 int라는 서로 다른 자료형을 비교해서 게임을 진행시키는 것이 어색하다라는 이유에서였습니다.

마찬가지로 Position도 TryCount와 같은 방식으로 만들었는데, 이는 어차피 VO에서 equals를 재정의해놓고 정작 int형을 파라미터로 받아서 같은지 판별하는 isSameWith 메소드를 사용하고 있다는 점에서 불필요한 메소드 추가라고 느꼈기 때문입니다. 그래서 TryCount와 Position 모두 VO로 만들면서 재정의한 equals를 적극 사용하는 방식으로 바꾸어보았습니다.

그리고 이렇게 되면 값을 증가시킬때도 생성자를 호출해야 하는 등, 객체를 생성하는 로직이 여러가지가 되기 때문에 불변 객체들을 생성할 때 생성자는 private으로 잠구고 정적 팩토리 메소드를 사용해서 적절한 네이밍을 붙여서 사용하도록 했습니다.

추가로 우승자를 구하는 로직에서 비교하는 메소드가 필요하기 때문에(기존에는 mapToInt를 활용했습니다.) Position을 Comparable의 구현체로 만들어서 compareTo 메소드를 재정의해주었습니다.

### view나 util 클래스들의 생성자 접근 제한
InputView, OutputView, ErrorMessages 등과 같이 모든 필드와 메소드가 static한 클래스의 경우 생성자 호출을 할 수 없도록 기본 생성자를 private으로 제한시켰습니다.

jjj0611 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -12,42 +15,55 @@
         OutputView.printStartMessage();
-        while (tryCount.isNotSame(currentTryCount++)) {
+        while (isGameContinuable(tryCount, currentTryCount)) {
             cars.moveAll(RandomNumberGenerator.fromBounds(RANDOM_LOWER_BOUND, RANDOM_UPPER_BOUND));
```
이 부분에 대한 리뷰를 놓쳤었네요
이름이 RandomNumberGenerator가 맞을까요

Ohzzi replied:
이 부분은 어떤 의도로 말씀해주시는지 잘 모르겠습니다...!

jjj0611 replied:
앗 요건 잘못 코멘트를 남겼네요.
while 문을 도는 동안에 계속 새로운 RandomNumberGenerator를 만들어낼 필요가 있을까요?

jjj0611 commented on src/main/java/racingcar/model/value/TryCount.java:
```diff
@@ -7,13 +7,21 @@
-    public TryCount(String countString) {
-        this.tryCount = convertStringToInt(countString);
-        validate(tryCount);
+    private TryCount(int tryCount) {
```
이 객체에 대해서 다시 생각해볼까요?
2개의 TryCount를 생성해서 비교를 하고 있는데, 조금 더 좋은 방법은 없을지 고민해보면 좋을 것 같아요!

만약 오찌는 StopWatch를 만들면 어떻게 만들 것 같나요?
StopWatch start = new StopWatch();
StopWatch end = new StopWatch();

의 방법과

StopWatch stopWatch = new StopWatch();
stopWatch.start();
stopWatch.end();

의 방법 중 어떤 것이 더 좋을까요?

TryCount가 지금 currentTryCount와 tryCount 두개가 만들어져서 항상 비교되어야 하는데,
시도 횟수를 하나의 객체로서 만들어볼 수 있지 않을까요?

Ohzzi replied:
하나의 객체를 만들어서 그 안에서 값을 조정하는게 더 깔끔해보이네요...! 그런데 이런 경우면 불변 객체로 만들어서 값을 올릴때 재할당을 시켜주는 방식을 채택하지 않고 객체 내부에 멤버 변수로 현재 시도값과 목표치인 시도횟수값을 두어서 비교하는 로직을 짜야겠네요??

jjj0611 commented on src/main/java/racingcar/model/value/TryCount.java:
```diff
@@ -7,13 +7,21 @@
-        return this.tryCount != tryCount;
+    public static TryCount fromString(String countString) {
+        return new TryCount(convertStringToInt(countString));
+    }
```
좋은 방법을 사용해주셨네요 :) 👍

jjj0611 commented on src/main/java/racingcar/model/value/TryCount.java:
```diff
@@ -7,13 +7,21 @@
+
+    public static TryCount initialize() {
+        return new TryCount(0);
+    }
```
이름이 조금 변경되어야 하지 않을까요?
TryCount를 생성하는 입장에서 initialize라는 것이 무엇을 의미하는지 알려면 한뎁스 더 들어와야할 것 같아요!

jjj0611 commented on src/main/java/racingcar/view/InputView.java:
```diff
@@ -7,6 +7,10 @@
 import racingcar.message.NoticeMessages;
 
 public class InputView {
+    private InputView() {
+
+    }
```
코멘트 없이도 잘 변경해주셨네요 👍

jjj0611 commented on src/main/java/racingcar/model/car/CarNameAndPosition.java:
```diff
@@ -3,13 +3,13 @@
  * Car 클래스의 Dto 클래스
  */
-public class CarStatus {
+public class CarNameAndPosition {
```
Status에서 이름을 변경해주셨네요!
만약 자동차의 이름, 포지션에 이어서 자동차에 팀의 개념이 부과 되면
CarNameAndPositionAndTeam이 되겠죠...?

Ohzzi replied:
마땅한 네이밍이 생각이 안나는데 어차피 담고 있는 정보가 무엇인가를 알려주는게 중요할 것 같아서 담고 있는 정보를 네이밍에 담았어요! 혹시 너무 불필요하게 길어지는걸까요...?

jjj0611 replied:
네 그럼요!
해당 Dto가 필드를 5개를 가지게 되면, 10개를 가지게 되면 어떻게 될까요....?

jjj0611 commented on src/main/java/racingcar/model/car/CarNameAndPosition.java:
```diff
@@ -3,13 +3,13 @@
-        this.name = car.getName();
-        this.position = car.getPosition();
+    public CarNameAndPosition(Car car) {
+        this.name = car.getName().toString();
+        this.position = car.getPosition().toInt();
```
car.getName()했을 때 String을 바로 반환해줘도 되지 않을까요??

Ohzzi replied:
혹시 Name 객체를 사용하는 로직이 추가될 수도 있을까봐 Name 객체를 반환하도록 했습니다.
불필요한 방법일까요??

jjj0611 commented on src/main/java/racingcar/model/car/Cars.java:
```diff
@@ -11,14 +11,15 @@
+        validateDuplicatedName(separatedNames);
+        cars = separatedNames.stream()
+                .map(Car::new)
+                .collect(Collectors.toList());
```
자 이제 고민을 한 번만 더 해보면 좋을 것 같아요!
지금은 carNames라는 String 덩어리를 받아서 Cars의 생성시점에 직접 split을 하여 생성을 하고 있는데,
그렇기 때문에 Cars에 ","라는 view와 관련된 매직넘버가 존재하게 되는 것 같아요.
String carNames -> List<String> carNames로 변환하는 로직이 view에 들어가면 어떤가요?
중복 체크 등의 로직은 Cars의 역할이지만, String -> List<String>의 변환까지 Cars의 도메인인가 하는 고민을 한 번 해보면 좋을 것 같아요 :)

Ohzzi replied:
안그래도 이 부분이 view에 있어야 하냐 controller에 있어야 하냐로 데일리 미팅에서 얘기가 있었었는데, 그 때 생각해 보니까 도메인에 있는건 확실히 불필요한 것 같더라고요. 저는 view에서 있는게 나을 것 같아서 view로 옮기도록 하겠습니다!

jjj0611 commented on src/main/java/racingcar/model/value/Position.java:
```diff
@@ -2,23 +2,32 @@
-        this.position += distance;
+    public static Position create() {
+        return new Position(INITIAL_POSITION);
+    }
```
여기도 Position.create()가 어떤 것을 만드는지도 좀 더 명시적으로 하면 어떨까요?

jjj0611 commented on src/test/java/racingcar/model/car/WinnersTest.java:
```diff
@@ -6,11 +6,9 @@
-import racingcar.model.car.Winners;
 
-@SuppressWarnings("NonAsciiCharacters")
 class WinnersTest {
```
오찌는 Winners를 따로 만들어서 테스트에서는 car를 각각 진행시키고 우승자를 구했는데,
만약 자동차의 진행 조건이 4에서 5로 바뀌면 어떻게 될까요?
해당 테스트가 전면 수정되어야 할겁니다.

항상 그런 것은 아니지만, 변경에 대해서 유연한 구조는
변경 후에 테스트에도 변경사항이 적다는 것을 의미할 때도 있어요.

마지막으로 인터페이스를 잘 활용해서 좀 더 테스트에 유연한 구조를 만들어보면 어떨까요?

Ohzzi replied:
변경 후에 테스트에도 변경사항이 적도록 하는 구조가 유연한 구조다 메모하겠습니다!
인터페이스를 활용하는 테스트 방식에 대한 논의가 크루들 사이에서 많아서 고민해보긴 했었는데 한 번 적용해봐야겠네요!

jjj0611 replied:
변경이 없을 수는 없으나 최소화하는 방향으로 고민해보면 좋을 것 같아요 :)

jjj0611 reviewed (changes requested):
안녕하세요 오찌! 리뷰어 제이입니다.

어느덧 자동차 경주게임이 마무리가 되어 가네요.
점점 더 깔끔해지고 보기 좋은 코드가 되어가고 있는 것 같아요 :)

고민해볼만한 부분, 적용해볼만한 부분 코멘트 드렸으니 한 번 봐주세요!
얼마 안남았습니다. 마지막까지 화이팅 💪🏻

Ohzzi committed:
refactor: 이름 String을 콤마로 자르는 로직을 view로 이동

Ohzzi committed:
refactor: NumberGenerator 인터페이스를 만들어서 구현체 사용하도록 수정

RandomNumberGenerator를 NumberGenerator 인터페이스의 구현체로 만들어서 Car까지 전달되도록 하고, 테스트 시에는 특정 값을 반환하는 MovableNumberGenerator, UnMovableNumberGenerator를 사용하도록 수정

Ohzzi committed:
refactor: Position 객체의 정적 팩토리 메소드 명 변경

출발점의 위치인 0을 기준으로 생성하므로 create -> fromStartLine으로 변경

Ohzzi committed:
refactor: 게임 진행 조건 로직 수정

TryCount -> GameRound로 바꾸고 처음 생성시에 게임을 진행할 totalRound 멤버 변수를 초기화 하면서 생성하고, 객체 안에 자체적으로 continuable 메소드를 만들어서 게임 진행 가능 여부를 판단하는 로직으로 수정

Ohzzi commented:
피드백 내용중에서 제가 이해하고 공감한 내용들에 대해서 수정해보았습니다!

1. String을 List<String> 형태로 자르는 로직을 Cars(model) -> view로 이동
2. NumberGenerator 인터페이스를 만들어 실제 프로덕션에서는 랜덤을, 테스트에서는 특정 값을 구현체로 전달해주도록 수정
    - 테스트시에 CsvSource나 ValueSource로는 테스트 되지 않아서 MethodSource를 사용했습니다.
    - RandomNumberGenerator의 이름에서 어떤 부분을 지적하시는지는 잘 모르겠어서 수정하지 않았습니다.
3. Position 객체를 initialize()로 생성하지 않고, 시작 시에는 출발선에서 시작할테니 fromStartLine() 으로 팩토리 메소드 명 변경
4. TryCount -> GameRound로 변경하고 게임 진행이 가능한지 비교 로직을 GameRound 안에 담아서 사용

추가적으로 피드백 해주실만한 내용 있으면 부탁드립니다! :)

jjj0611 commented on src/main/java/racingcar/util/MovableNumberGenerator.java:
```diff
@@ -0,0 +1,8 @@
+package racingcar.util;
+
+public class MovableNumberGenerator implements NumberGenerator {
```
src/main에 있는게 맞을까요?

jjj0611 commented on src/main/java/racingcar/util/UnMovableNumberGenerator.java:
```diff
@@ -0,0 +1,8 @@
+package racingcar.util;
+
+public class UnMovableNumberGenerator implements NumberGenerator {
```
역시나 src/main에 있는게 맞을까요?
더불어서 0을 리턴할 것이 아니라 엣지 케이스로 테스트하면 좋을 것 같아요 :)

jjj0611 reviewed (approved):
안녕하세요 오찌! 리뷰어 제이입니다.

자동차 경주게임 리팩토링 완료해주신 것 잘 보았어요.
추가로 코멘트 몇 개 드렸으니 확인 부탁드려요!
그동안의 리뷰가 도움이 많이 되었길 바랍니다!
미션 진행하시느라 고생하셨고, 코드는 이만 머지하겠습니다.
다음 미션도 화이팅 하세요!
